independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,72 @@
+package io.quarkus.bootstrap.runner;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.jar.JarFile;
+
+import io.smallrye.common.io.jar.JarFiles;
+
+public class JarFileReference {
+    // Guarded by an atomic reader counter that emulate the behaviour of a read/write lock.
+    // To enable virtual threads compatibility and avoid pinning it is not possible to use an explicit read/write lock
+    // because the jarFile access may happen inside a native call (for example triggered by the RunnerClassLoader)
+    // and then it is necessary to avoid blocking on it.
+    private final JarFile jarFile;
+
+    private final AtomicInteger readersCounter = new AtomicInteger(0);
+
+    private volatile boolean releasing = false;
+
+    private JarFileReference(JarFile jarFile) {
+        this.jarFile = jarFile;
+    }
+
+    JarFile readLockAcquire() {
+        readersCounter.incrementAndGet();
+        if (releasing) {
+            if (readersCounter.decrementAndGet() == 0) {
+                close();
+            }
+            return null;
+        }
+        return jarFile;
+    }
+
+    boolean readLockRelease() {
```

Comment:
A little Javadoc here which explains what the return value means would be good IMO.

Reply:
Done.